### DISCUSSION TITLE
WSL USB Manager: Windows GUI tool for WSL

### DISCUSSION THREAD

**Discussion by nickbeth** [dorssel/usbipd-win#926] [Show and tell]
Hey,
I've been using `usbipd-win` for the past year and I've been loving it so far; it has been crucially important for working on embedded systems on Windows, allowing me to avoid the horrendous STMicroelectronics official tooling and just use Linux and the `stlink-org/stlink` tools instead.

I've recently started learning Rust and wanted to build something with it, so why not make a GUI for `usbipd-win`, although one already exists?
The current GUI has 3 main issues in my opinion:
- Unnecessarily big: 190MB install size
- Slow to open: sometimes taking more than 10 seconds
- Unresponsive when in energy saving mode (CPU running at low clock speeds, works wonders for battery life)

~~I also needed an auto-attachment thing that persists reboots. It would start up when the user logs in, runs in the background and automatically attaches USB devices to WSL.~~

The GUI I'm presenting here fixes all of the above issue, being a 600KB executable that leverages the "old" Win32 API for fast startup times and responsiveness. It also has a native look & feel as an added bonus.
The UI is organized in a more user-friendly way, displaying all the available details of a given USB device on a panel on the right of the device list. I've also tried to keep the same separation as the `usbipd list` command, which shows separate lists for `Connected` and `Persisted` devices.
At the moment, only the basic operations are supported (bind/unbind/attach/detach). ~~Auto-attach is **not yet implemented**, but it's coming soon™️. I wanted to get the basics to work first, before continuing with the auto-attach feature.~~

Here's the link to the repo: https://github.com/nickbeth/wsl-usb-manager

Update: Auto-attach has been implemented, it leverages the `usbipd attach --auto-attach` command, which means auto-attach profiles are reset every time the application is restarted. Getting auto-attach profiles to persist reboot is harder than one may think as there is no real way to uniquely identify devices across reboots and plug/unplug cycles, and I've also realized I didn't really need that feature to begin with.

![WSL USB Manager](https://raw.githubusercontent.com/nickbeth/wsl-usb-manager/master/img/connected_devices.png)

**Comment by callimero** [comment]
Just want to say thanks! That's how a tool needs to be. 

Carsten

**Comment by JPHutchins** [comment]
Wow, this looks great!  I don't really have any issues with WSL USB Manager 5.3.0, but as someone who has been working on "packaging python into portable apps" lately, I understand all the pain that can be inflicted by developing native applications in Python!  It's not going to get better... so better to switch to Rust 🦀.

**Reply by callimero** [reply]
Hmm? It is rust?

**Comment by dorssel** [comment]
I've added you to the new Wiki page for GUI/IDE tools: https://github.com/dorssel/usbipd-win/wiki/Graphical-User-Interfaces.

**Reply by nickbeth** [reply]
Thanks!

**Comment by mstempin** [comment]
> Update: Auto-attach has been implemented, it leverages the `usbipd attach --auto-attach` command, which means auto-attach profiles are reset every time the application is restarted. Getting auto-attach profiles to persist reboot is harder than one may think as there is no real way to uniquely identify devices across reboots and plug/unplug cycles, and I've also realized I didn't really need that feature to begin with.

Do it like WSL-USB GUI's `Edit Custom Profile`: plain string or regex match on Bus ID, Description, and/or Instance ID.
<img width="322" height="271" alt="2025-09-27 22 05 51" src="https://github.com/user-attachments/assets/5fa62878-2266-470b-bec9-0393b0e17ec3" />

Unfortunately, not having auto-attach across reboots and plug/unplug cycles is a show-stopper for me: I have 5 devices to reattach at once. ~At least this can be done by relaunching the WSL-USB GUI, but it cannot be automated by putting it into the Windows startup because it is Python.~

**Comment by totszwai** [comment]
Don't download it, flagged as a virus, who the hell puts a raw exe as a release?!

https://github.com/nickbeth/wsl-usb-manager/issues/31

**Reply by JPHutchins** [reply]
> Don't download it, flagged as a virus, who the hell puts a raw exe as a release?!
> 
> 
> 
> https://github.com/nickbeth/wsl-usb-manager/issues/31

This is a Microsoft problem, not a real security issue. It's much easier for Microsoft to flag open source software than to actually investigate every application out there.

It is open source. I am a contributor. The author uses their real name on GitHub. There is no malicious software here.

**Reply by totszwai** [reply]
The point is that, nobody should be putting download links straight to a exe. Not to mention that repo has some kind of installer defined, the repo owner should just build it and release that instead.

**Reply by JPHutchins** [reply]
I do disagree. I love portable executables, and many projects have an option to be distributed in this way. Just one example of a long lived OSS project that has "portable" or "standalone" binaries available is PuTTY: 

https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html

The very real issue that you're raising is one of code signing. I will work with @nickbeth and Microsoft to see if they will sponsor signing for our project. The WSL and code signing teams have been responsive in the past, but it's a huge company with hard to move pieces. We have a ticket about signing here: https://github.com/nickbeth/wsl-usb-manager/issues/2

The reason to use installers is to comply with OS tooling and provide secure storage for non volatile user data. In the case of the WSL USB Manager, it would allow for us to save user settings (for auto attach) and debug logs across runs.

We would welcome assistance with testing and fine tuning the installer: https://github.com/nickbeth/wsl-usb-manager/tree/master/msi

**Reply by totszwai** [reply]
Sure, but you also need to be aware that these kind a thing happens more and more often these days.

https://www.malwarebytes.com/blog/threat-intel/2026/02/fake-7-zip-downloads-are-turning-home-pcs-into-proxy-nodes

Regardless, that's where my concern came from.
